File: services/typescript/fastify/src/routes/handlers/patch.ts
import type { RouteHandler } from "fastify";
import { type State, parseId, parsePatchedTask } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";
import type { PatchRoute } from "../types.js";

export function patchHandler(state: State): RouteHandler<PatchRoute> {
  return async (request, reply) => {
    const logger = getLogger();

    const id = parseId(request.params.id);
    if (!id.success) {
      logger.warn(
        { method: request.method, path: request.url },
        "Invalid path"
      );
      return sendError(reply, AppErrors.InvalidPath);
    }

    const patchedTask = parsePatchedTask(request.body);
    if (!patchedTask.success) {
      logger.warn(
        { method: request.method, path: request.url },
        "Invalid body JSON"
      );
      return sendError(reply, AppErrors.InvalidJsonBody);
    }

    const task = state.tasks.get(id.data);
    if (!task) {
      logger.warn(
        { id: id.data, method: "PATCH", path: request.url },
        "Task not found"
      );
      return sendError(reply, AppErrors.TaskNotFound);
    }

    const previousOperation = task.operation;
    task.operation = patchedTask.data.operation;

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        from_operation: previousOperation.toString().toLowerCase(),
        to_operation: task.operation.toString().toLowerCase(),
        method: "PATCH"
      },
      "Patched task"
    );

    return reply.status(200).send(task);
  };
}

File: services/python/fastapi/src/routes/handlers/patch.py
import structlog
from uuid import UUID
from fastapi import Request
from fastapi.responses import JSONResponse

from app.state import AppState
from app.task import PatchedTask, Task
from app.operation import Operation
from routes.errors import send_error, AppErrors


async def patch_handler(
    request: Request, id: UUID, patched_task: PatchedTask
) -> JSONResponse | Task:
    logger = structlog.get_logger()

    state: AppState = request.app.state.app_state

    async with request.app.state.tasks_lock:
        task: Task | None = state.tasks.get(id)

        if task is None:
            logger.warn(
                "Task not found", id=str(id), method="PATCH", path=request.url.path
            )
            return send_error(AppErrors.TaskNotFound)

        previous_operation: Operation = task.operation

        state.tasks[id].operation = patched_task.operation

    logger.info(
        "Patched task",
        id=str(id),
        secret=len(task.secret),
        from_operation=previous_operation.lower(),
        to_operation=patched_task.operation.lower(),
        method="PATCH",
    )

    return task
